### PR TITLE
Add cypress.io for client end to end tests

### PR DESCRIPTION
**Pull request by Bouncey**
- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] My changes do not use shortened URLs or affiliate links.

This PR introduces initial e2e tests for the client.

This is the first step in a process that will eventually move the `curriculum/test/test-challenges.js` tests in to the browser.

### CODE REVIEW

**Review by raisedadead — COMMENTED** [review]
Would it make sense to move cypress to root?

**Comment by ojeytonwilliams** [comment]
@Bouncey this looks great!  One thing I don't get though, `npm run e2e:ci` in the client, what is that supposed to do?  If I run it, it builds the client, serves it and then just sits there.  Even if there are errors that get flagged up by `npm run e2e`.  Apologies if I'm missing something obvious!

**Comment by Bouncey** [comment]
@ojeytonwilliams I'm not quite sure what you mean. Maybe you were on an older version of the branch when you checked it out. Everything has moved to the root now.

**Review by raisedadead — APPROVED** [review]

File: cypress/run-e2e.sh [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  # Find any node processes running from within the client dir
+  local hanging_client_processes=$(ps aux | grep -v grep | grep client/node_modules | awk '{print $2}')
+  local hanging_api_processes=$(ps aux | grep -v grep | grep api-server/node_modules | awk '{print $2}')
+
+  # Send kill signal to the processes
+  if [ ${#hanging_api_processes} -gt "0" ];  then
+    kill -9 $hanging_api_processes &>/dev/null
+  fi
+  if [ ${#hanging_client_processes} -gt "0" ]; then
+    kill -9 $ hanging_client_processes &>/dev/null
```
Comment (ojeytonwilliams) [inline cypress/run-e2e.sh]:
```suggestion
    kill -9 $hanging_client_processes &>/dev/null
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
The problem I encountered before has gone away, so that's all good.

I found a few minor things:

File: cypress/run-e2e.sh [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  cd client
+  npm run stand-alone &
+  gastby_pid=$!
+
+
+  cd ../api-server
+  npm start &
+  api_pid=$!
+
+  cypress_cmd='cypress:run'
```
Comment (ojeytonwilliams) [inline cypress/run-e2e.sh]:
```suggestion
  cypress_cmd='cypress:run-dev'
```
To make use of the new npm script.

**Comment by ojeytonwilliams** [comment]
I also noticed that `node production-start.js` doesn't get killed in the clean up.  Is it possible to use the api-server in development mode?  If not, that should be cleaned up, too.

**Comment by Bouncey** [comment]
@ojeytonwilliams, sorry, I should have communicated this. CircleCI cannot build the production version of the client because the environment does not have enough memory. This is something we found with netlify/travis/appveyour and others, and is why we use other methods to build the production client.

I left the code there for if/when the build time memory issues are resolved.

As to your hanging process, I might need you to look at it as I cannot reproduce on my machine.

**Comment by ojeytonwilliams** [comment]
I see, thanks!  Is it worth setting it up so that you can test the production version locally, even if you can't do it in ci?  It might be nice to catch the slight differences between the static site and the development mode (hydration, mostly).  I don't know enough about cypress and e2e testing to know if that's worth doing, though.

Anyway, the hanging process.  When I run `npm run e2e` from the root that starts a bunch of processes, but only the initial `npm` processes and the shell scripts they spawn get killed by

```sh
  kill -9 $gastby_pid $api_pid &>/dev/null
```
the problem is that the `node production-start.js` process remains after that.  I can pop the fix in a commit, if you like - bit awkward doing it via Github's suggestions.

**Comment by raisedadead** [comment]
I think we could use Azure Pipelines (which we already do for production builds) and add a stage for the e2e tests.

**Comment by ahmadabdolsaheb** [comment]
@Bouncey, any updates on this : )

**Comment by raisedadead** [comment]
Stuart made this PR in an effort to guide us on how we can integrate Cypress into our setup. This is more like 'Hello World' for us, and we are meant to pick it up from here. :)

**Comment by ojeytonwilliams** [comment]
I have brought things up to date.  Next step, see if Github actions has enough memory (they have access to 7GB, apparently) to handle building the client.

**Comment by raisedadead** [comment]
> ... Github actions has enough memory (they have access to 7GB, apparently) to handle building the client.

You want to use Azure Pipelines, which we have been using to build the client? If you need access to the org or want me to chime instead, I would be happy to help.

**Comment by ojeytonwilliams** [comment]
@raisedadead not sure, honestly.  I saw your earlier comment about Azure, but I figured I'd try out Actions first because I'm more familiar with them.

Also, there's already an action we can use: https://github.com/cypress-io/github-action which should make things easier.

What do you reckon?  Is it worth giving Actions a go?

**Comment by raisedadead** [comment]
Yes, sure. I am happy to use GitHub actions. If we ran into limitations, we could switch. Anything is better than nothing at this point. 

And 100% agree that end-user tests are going to be more critical than the unit tests.

**Review by raisedadead — APPROVED** [review]
This looks like a good starting point. I noticed that you removed the Circle CI config, which is fine. I have created this https://github.com/freeCodeCamp/freeCodeCamp/issues/38611

We may need something that will run the entire freeCodeCamp curriculum.

**Comment by ojeytonwilliams** [comment]
@raisedadead since Actions are working well, Circle CI seemed redundant.

**Pull request merged**